Fix for ids not being escaped in paths in the Simple Rest Data Provider

ckhsponge opened the pull request:
## Problem

Records may have ids with characters that break the Simple REST Data Provider paths if not escaped. 
For example, a record with id "Post#123" has a getOne path of '/posts/Post#123' which is not desired.
IDs like this can occur when using single table inheritance with DynamoDB or with other document data stores.

## Solution

The solution is to escape IDs in paths for getOne, update, updateMany, delete and deleteMany. An example path with escaped id is '/posts/Post%23123'.

Closes #10207

## How To Test

Perform getOne, update, updateMany, delete and deleteMany with records that have escape-requiring ids e.g. 'Post#123'. 

## Additional Checks

- [ x ] The PR targets `master` for a bugfix, or `next` for a feature
- [ x ] The PR includes **unit tests** (if not possible, describe why)
- [ ] The PR includes one or several **stories** (if not possible, describe why)
- [ x ] The **documentation** is up to date

Also, please make sure to read the [contributing guidelines](https://github.com/marmelab/react-admin#contributing).

fzaninotto commented on packages/ra-data-simple-rest/src/index.ts:
```diff
@@ -156,10 +156,13 @@ export default (
-                    body: JSON.stringify(params.data),
-                })
+                httpClient(
+                    `${apiUrl}/${resource}/${encodeURIComponent(typeof id === 'object' ? id.id : String(id))}`,
```
I don't understand why you test that `id` may be an object. The interface doesn't allow for that. Same for `deleteMany`.

ckhsponge replied:
An existing deleteMany test had that as input so I assumed that was the format. Fixing.

fzaninotto commented on packages/ra-data-simple-rest/src/index.spec.ts:
```diff
@@ -68,6 +68,65 @@ describe('Data Simple REST Client', () => {
     });
+    describe('getOne', () => {
+        it('should allow numeric id in path', async () => {
+            const httpClient = jest.fn().mockResolvedValue({ id: 'Post#123' });
```
This may cause a warning as the requested `id` and the returned `id` aren't the same.

ckhsponge replied:
Fixing.

fzaninotto commented on packages/ra-data-simple-rest/src/index.spec.ts:
```diff
@@ -68,6 +68,65 @@ describe('Data Simple REST Client', () => {
+        it('should escape id in path', async () => {
+            const httpClient = jest
+                .fn()
+                .mockResolvedValue({ json: { id: 'Post#123' } });
```
This isn't a valid `updateMany` response and it will cause warnings.

ckhsponge replied:
Fixing.

fzaninotto commented on packages/ra-data-simple-rest/src/index.spec.ts:
```diff
@@ -120,5 +193,20 @@ describe('Data Simple REST Client', () => {
+        it('should escape id in path', async () => {
+            const httpClient = jest
+                .fn()
+                .mockResolvedValue({ json: { id: 'Post#123' } });
```
This isn't a valid `updateMany` response

ckhsponge replied:
Fixing.

fzaninotto reviewed (changes requested):
Thanks for your PR! I found a few problems to be fixed before we can merge it.

ckhsponge commented:
Thanks for reviewing my PR!

I think an existing deleteMany test had an incorrect input which made me think objects were allowed as inputs there and for updateMany. I fixed that test and mine and removed that check from the data provider.

fzaninotto commented:
Thanks!
